policy/nomad/parser.go

```diff
@@ -4,107 +4,161 @@ import (
+	if p.Enabled == nil {
+		to.Enabled = true
+	} else {
+		to.Enabled = *p.Enabled
 	}
 
-	canonicalizePolicy(p, &to)
+	// Parse evaluation_interval as time.Duration.
+	// Ignore error since we assume policy has been validated.
+	if eval, ok := p.Policy[keyEvaluationInterval].(string); ok {
+		to.EvaluationInterval, _ = time.ParseDuration(eval)
+	}
 
-	return to, nil
+	return to
 }
 
+// parseStrategy parses the content of the strategy block from a policy.
+//
+// It provides best-effort parsing and will return `nil` in case of errors.
+//
+//  scaling {
+//    policy {
+//      strategy = {
+//      +-------------------+
+//      | name = "strategy" |
+//      | config = {        |
+//      |   key = "value"   |
+//      | }                 |
+//      +-------------------+
+//      }
+//    }
+//  }
 func parseStrategy(s interface{}) *policy.Strategy {
-	strategyMap := s.([]interface{})[0].(map[string]interface{})
-	configMap := strategyMap["config"].([]interface{})[0].(map[string]interface{})
-	configMapString := make(map[string]string)
-	for k, v := range configMap {
-		configMapString[k] = fmt.Sprintf("%v", v)
+	if s == nil {
+		return nil
+	}
+
+	strategyMap := parseBlock(s)
+	if strategyMap == nil {
+		return nil
+	}
+
+	var configMapString map[string]string
+	configMap := parseBlock(strategyMap["config"])
+
+	if configMap != nil {
+		configMapString = make(map[string]string)
+		for k, v := range configMap {
+			configMapString[k] = fmt.Sprintf("%v", v)
+		}
 	}
 
+	// Ignore ok, but we need _ to avoid panics.
+	name, _ := strategyMap["name"].(string)
+
 	return &policy.Strategy{
-		Name:   strategyMap["name"].(string),
+		Name:   name,
 		Config: configMapString,
 	}
 }
 
-func parseTarget(t interface{}) *policy.Target {
-	if t == nil {
-		return &policy.Target{}
+// parseTarget parses the content of the target block from a policy and
+// enhances it with values defined in Target as well. Values in target.config
+// takes precedence over values in Target.
+//
+// It provides best-effort parsing and will return `nil` in case of errors.
+//
+//  scaling {
+//    policy {
+//      target = {
+//      +-----------------+
+//      | name = "target" |
+//      | config = {      |
+//      |   key = "value" |
+//      | }               |
+//      +-----------------+
+//      }
+//    }
+//  }
+func parseTarget(targetBlock interface{}, targetAttr map[string]string) *policy.Target {
+	if targetBlock == nil && targetAttr == nil {
+		return nil
 	}
 
-	targetMap := t.([]interface{})[0].(map[string]interface{})
+	targetMap := parseBlock(targetBlock)
 	if targetMap == nil {
-		return &policy.Target{}
+		return nil
 	}
 
 	var configMapString map[string]string
-	if targetMap["config"] != nil {
-		configMap := targetMap["config"].([]interface{})[0].(map[string]interface{})
+	configMap := parseBlock(targetMap["config"])
+
+	if configMap != nil {
 		configMapString = make(map[string]string)
+
+		for k, v := range targetAttr {
```

Comment:
on my phone, can't trace this back, but wanted to make sure that this is only expected to be set if "config" existed

Reply:
Yes, this moving values from `targetMap["config"]` (which is a `map[string]interface{}`) into `configMapString` (which is `map[string]string`).

So if `"config"` is not set we don't need to do anything.